### ISSUE TITLE
vak learncurve crashes trying to load data out of the expected range 

### ISSUE THREAD

**Issue by yardencsGitHub** [NickleDave/vak#217]
Running with a training set of 240 sec I encountered the following crash.
If I'm reading it correctly, this means that the dataloader tried to get data from beyond the bounds of a file. 

Exception has occurred: IndexError
Caught IndexError in DataLoader worker process 2.
Original Traceback (most recent call last):
  File "C:\Anaconda3\envs\tweetyenv_alpha\lib\site-packages\torch\utils\data\_utils\worker.py", line 178, in _worker_loop
    data = fetcher.fetch(index)
  File "C:\Anaconda3\envs\tweetyenv_alpha\lib\site-packages\torch\utils\data\_utils\fetch.py", line 44, in fetch
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "C:\Anaconda3\envs\tweetyenv_alpha\lib\site-packages\torch\utils\data\_utils\fetch.py", line 44, in <listcomp>
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "c:\users\yarden cohen\repos\vak_split_br\src\vak\datasets\window_dataset.py", line 210, in __getitem__
    window, labelvec = self.__get_window_labelvec(idx)
  File "c:\users\yarden cohen\repos\vak_split_br\src\vak\datasets\window_dataset.py", line 186, in __get_window_labelvec
    spect_id = self.spect_id_vector[x_ind]
IndexError: index 90551 is out of bounds for axis 0 with size 88889
  File "C:\Users\Yarden Cohen\repos\vak_split_br\src\vak\engine\model.py", line 109, in _train
    for ind, batch in enumerate(progress_bar):
  File "C:\Users\Yarden Cohen\repos\vak_split_br\src\vak\engine\model.py", line 415, in fit
    ckpt_step)
  File "C:\Users\Yarden Cohen\repos\vak_split_br\src\vak\core\train.py", line 289, in train
    device=device)
  File "C:\Users\Yarden Cohen\repos\vak_split_br\src\vak\core\learncurve.py", line 268, in learning_curve
    **window_dataset_kwargs
  File "C:\Users\Yarden Cohen\repos\vak_split_br\src\vak\cli\learncurve.py", line 74, in learning_curve
    logger=logger,
  File "C:\Users\Yarden Cohen\repos\vak_split_br\src\vak\cli\cli.py", line 31, in cli
    learning_curve(toml_path=config_file)
  File "C:\Users\Yarden Cohen\repos\vak_split_br\src\vak\__main__.py", line 44, in main
    config_file=args.configfile)
  File "C:\Users\Yarden Cohen\repos\vak_split_br\src\vak\__main__.py", line 48, in <module>
    main()

**Comment by NickleDave** [comment]
Thank you for letting me know

From the path
```
File "c:\users\yarden cohen\repos\vak_split_br\src\vak\datasets\window_dataset.py", line 186, in _get_window_labelvec
```
(i.e. the `vak_split_br`) I'm wondering if this is your fork?
Does that mean you got this bug after making changes to the code I don't have?
Did you hit the bug when you were testing the `fix-split-brute-force` branch?

**Comment by yardencsGitHub** [comment]
Yes.

It is in my branch.

**Comment by NickleDave** [comment]
Okay, in that case, we're not sure if it's related to the new code or something that was pre-existing

Are you able to reproduce the bug reliably every time you run?

What happens if you check back out master and then run? Do you still get the bug?

One option would be to go ahead and start the PR -- I can begin code review and that might help us spot anything

**Comment by yardencsGitHub** [comment]
I’ll have time tonight or tomorrow to check these

**Comment by NickleDave** [comment]
I think this is fixed by #219 
but capturing some back and forth from email here for future reference:

from @yardencsGitHub 
> When the __get_window_labelvec function runs properly:
> self.x_inds is as it should be. Values ranging from 0 to the max index in self.spect_inds_vector
> 
> Also, the vector self.spect_id_vector has values starting from 0 (in the case I’m looking at now 0..10)
> 
> When the __get_window_labelvec function fails:
> self.x_inds has values in an offset. Values ranging from some number that I did not identify (9943) to a max value that exceeds the length of the self.spect_id_vector
> 
> Subtracting the min value of self.x_inds from its max value I found that it is 1 sec (370 bins) shorter than the actual duration of the training set.
> 
> Also, the vector self.spect_id_vector has values starting from 1 and not 0.

and then @NickleDave replied back

> I'm thinking it has something to do with what happens when x_inds gets cropped from the front.
> 
> That would explain why x_inds starts from some number greater than zero, because the inds at the beginning got cropped off, and could explain why spect_id_vector starts from 1, because also the bins from the first spectrogram got cropped off.
> 
> But I don't see where error in my logic for cropping is.
> This would explain why I didn't see the error too (besides the lack of tests for this explicit case, cropping from the front) 
> Probably it was always okay to crop from the back for the BF song since we were just throwing out the last 88 time bins.
 
and then finally @yardencsGitHub replied back

> I hope I caught the bug in WindowDataset.
> 
> I think you were right and it is the cropping in the beginning of the indices vector.
> 
> If x_inds points into positions in the training dataset then line 491:
> `x_inds = x_inds[x_inds != WindowDataset.INVALID_WINDOW_VAL]`
> 
> crops x_inds but if it crops the beginning it will start with numbers larger than 0 and point to positions outside spect_id_vector.
> 
> I changed to:
> ```
> x_inds = x_inds[x_inds != WindowDataset.INVALID_WINDOW_VAL]
> x_inds = x_inds - x_inds[0]
> ```
> 
> if x_inds indeed points to relative positions in spect_id_vector and not absolute positions in spectrograms and label vectors then this should work.
